### PR TITLE
MLIR interface lib

### PR DESCRIPTION
**Pull request by mbezuljTT**
Merging to the internal feature branch, required reviewers no actions required from your side

### Problem description
Defining a static library c++ `ttnn_mlir_interface.a` as a wrapper around the op constraints code. this is what mlir is to consume. Not final version, close enough to start integration on the mlir side. 

Missing matmul test - to followup on the separate PR.

### What's changed

`ttnn/cpp/ttnn/mlir_interface_api.hpp` is library header.
`tests/ttnn/unit_tests/gtests/test_mlir_interface_lib.cpp` is wrapper library test, isn't focused on e2e testing against silicon nor dispatcher.

### CODE REVIEW

File: ttnn/CMakeLists.txt [hunk 1]
```diff
@@ -449,3 +449,74 @@ set_target_properties(ttnn PROPERTIES
...
+    # ${CMAKE_CURRENT_SOURCE_DIR}/cpp/ttnn/tensor/types.cpp                     # Shape
+    # ${CMAKE_CURRENT_SOURCE_DIR}/../tt_metal/impl/buffers/buffer.cpp         # Buffer
+    # ${CMAKE_CURRENT_SOURCE_DIR}/../tt_metal/tt_metal.cpp                   # tt::tt_metal::detail::AllocateBuffer
+    # ${CMAKE_CURRENT_SOURCE_DIR}/../tt_metal/impl/allocator/allocator.cpp   # tt::tt_metal::Device::logical_core_from_bank_id
+    # ${CMAKE_CURRENT_SOURCE_DIR}/../tt_metal/impl/device/device.cpp              # tt::tt_metal::Device::dram_channel_from_bank_id
+    # ${CMAKE_CURRENT_SOURCE_DIR}/../tt_metal/llrt/tt_cluster.cpp      # tt::Cluster::instance()
+    # ${CMAKE_CURRENT_SOURCE_DIR}/../tt_metal/llrt/llrt.cpp               # tt::llrt::write_hex_vec_to_core
+    # ${CMAKE_CURRENT_SOURCE_DIR}/../tt_metal/impl/device/device_pool.cpp # tt::DevicePool::_ins
+    # ${CMAKE_CURRENT_SOURCE_DIR}/../tt_metal/impl/program/program.cpp # tt::tt_metal::Program::compile
+    # this list is too long, linking ttnn.so for now
```
Comment (mbezuljTT) [inline ttnn/CMakeLists.txt:475]: I originally hoped we could get away from this, but leaving this problem for later.

**Review by derdeljanTT — APPROVED** [review]

File: ttnn/cpp/ttnn/types_wrapper.cpp [hunk 2]
```diff
@@ -0,0 +1,418 @@
...
+
+    std::optional<tt::tt_metal::MemoryConfig> to_memory_config(const mlir_interface::memory_config_tuple& memory_config_tuple) {
+        return to_memory_config(
+            std::get<0>(memory_config_tuple),
+            std::get<1>(memory_config_tuple),
+            std::get<std::optional<mlir_interface::shard_spec_tuple>>(memory_config_tuple));
+    }
+
+    // MatmulMultiCoreReuseProgramConfig wrapper
+    std::optional<ttnn::operations::matmul::MatmulMultiCoreReuseProgramConfig> to_program_config(
```
Comment (ajakovljevicTT) [inline ttnn/cpp/ttnn/types_wrapper.cpp:319]: nit: Should this maybe be named something like `to_multi_core_program_config`, to be in line with other the other two methods?

File: ttnn/cpp/ttnn/types_wrapper.hpp [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    std::optional<tt::tt_metal::BufferType> to_buffer_type(const std::string& buffer_type_str);
+    std::optional<tt::tt_metal::ShardOrientation> to_shard_orientation(const std::string& shard_str);
+    std::optional<ttnn::operations::unary::UnaryOpType> to_unary_op_type(const std::string& unary_op_type_str);
+} // namespace ttnn::str_wrapper
+
+namespace ttnn::tuple_wrapper
+{
+    std::optional<tt::tt_metal::ShardSpec> to_shard_spec(const mlir_interface::shard_spec_tuple& shard_spec_tuple);
+    std::optional<tt::tt_metal::MemoryConfig> to_memory_config(const mlir_interface::memory_config_tuple& memory_config_tuple);
+    std::optional<ttnn::operations::matmul::MatmulMultiCoreReuseProgramConfig> to_program_config(const mlir_interface::matmul_multicore_reuse_config_tuple& program_config_tuple);
```
Comment (ajakovljevicTT) [inline ttnn/cpp/ttnn/types_wrapper.hpp:62]: nit: Should this maybe be named something like to_multi_core_program_config, to be in line with other the other two methods?

**Review by ajakovljevicTT — APPROVED** [review]

File: ttnn/cpp/ttnn/types_wrapper.hpp (hunk 3 above)
Comment (mbezuljTT) [reply ttnn/cpp/ttnn/types_wrapper.hpp]: renamed per offline conversation

**Pull request merged**